File: python/crypto-square/crypto_square.py
import math


def cipher_text(plain_text):
    return ' '.join([''.join([segment[i] if i < len(segment) else ' ' for segment in get_plain_text_segments(plain_text)]) for i in range(get_square_size(plain_text))])


def get_square_size(plain_text):
    return math.ceil(math.sqrt(len(get_normalized_plain_text(plain_text))))


def get_normalized_plain_text(plain_text):
    return ''.join(filter(str.isalnum, plain_text.lower()))


def get_plain_text_segments(plain_text):
    normalized_plain_text = get_normalized_plain_text(plain_text)
    square_size = get_square_size(plain_text)
    return [normalized_plain_text[i:i + square_size] for i in range(0, len(normalized_plain_text), square_size)]

File: java/crypto-square/src/main/java/CryptoSquare.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class CryptoSquare {
  String s;

  CryptoSquare(String s) {
    this.s = s;
  }

  String getNormalizedPlaintext() {
    return s.toLowerCase()
        .chars()
        .filter(Character::isLetterOrDigit)
        .mapToObj(c -> (char) c)
        .map(String::valueOf)
        .collect(Collectors.joining());
  }

  int getSquareSize() {
    return (int) Math.ceil(Math.sqrt(getNormalizedPlaintext().length()));
  }

  List<String> getPlaintextSegments() {
    List<String> segments = new ArrayList<>();
    String normalizedPlaintext = getNormalizedPlaintext();
    int squareSize = getSquareSize();
    for (int i = 0; i < normalizedPlaintext.length(); i += squareSize) {
      segments.add(
          normalizedPlaintext.substring(i, Math.min(i + squareSize, normalizedPlaintext.length())));
    }

    return segments;
  }

  String getCiphertext() {
    StringBuilder result = new StringBuilder();
    int squareSize = getSquareSize();
    List<String> segments = getPlaintextSegments();
    for (int i = 0; i < squareSize; ++i) {
      if (i != 0) {
        result.append(" ");
      }

      for (String segment : segments) {
        result.append((i < segment.length()) ? segment.charAt(i) : " ");
      }
    }

    return result.toString();
  }

  String getNormalizedCipherText() {
    StringBuilder result = new StringBuilder();
    int squareSize = getSquareSize();
    List<String> segments = getPlaintextSegments();
    for (int i = 0; i < squareSize; ++i) {
      if (!result.isEmpty()) {
        result.append(" ");
      }

      for (String segment : segments) {
        if (i < segment.length()) {
          result.append(segment.charAt(i));
        }
      }
    }

    return result.toString();
  }
}
